crates/db_views/src/post/post_view.rs

```diff
@@ -141,179 +154,111 @@ impl PostView {
+  pub async fn prefetch_cursor_before_data(
     pool: &mut DbPool<'_>,
-    local_user: Option<&LocalUser>,
-  ) -> Result<PaginationCursorData, Error> {
-    let err_msg = || Error::QueryBuilderError("Could not parse pagination token".into());
-    let post_id = PostId(
-      self
-        .0
-        .get(1..)
-        .and_then(|e| i32::from_str_radix(e, 16).ok())
-        .ok_or_else(err_msg)?,
-    );
-    let post = Post::read(pool, post_id).await?;
-    let post_actions = PostActionsCursor::read(pool, post_id, local_user.person_id()).await?;
-
-    Ok(PaginationCursorData { post, post_actions })
+    my_local_user: Option<&'_ LocalUser>,
+    page_back: Option<bool>,
+    limit: Option<i64>,
+  ) -> LemmyResult<Option<Post>> {
+    let conn = &mut get_conn(pool).await?;
+    let limit = limit_fetch(limit)?;
+    // first get one page for the most popular community to get an upper bound for the page end for
+    // the real query. the reason this is needed is that when fetching posts for a single
+    // community PostgreSQL can optimize the query to use an index on e.g. (=, >=, >=, >=) and
+    // fetch only LIMIT rows but for the followed-communities query it has to query the index on
+    // (IN, >=, >=, >=) which it currently can't do at all (as of PG 16). see the discussion
+    // here: https://github.com/LemmyNet/lemmy/issues/2877#issuecomment-1673597190
+    //
+    // the results are correct no matter which community we fetch these for, since it basically
+    // covers the "worst case" of the whole page consisting of posts from one community
+    // but using the largest community decreases the pagination-frame so make the real query more
+    let self_person_id = my_local_user
+      .person_id()
+      .ok_or(LemmyErrorType::CouldntParsePaginationToken)?;
+
+    let largest_subscribed = community_actions::table
+      .filter(community_actions::followed.is_not_null())
+      .filter(community_actions::person_id.eq(self_person_id))
+      .inner_join(community::table.on(community::id.eq(community_actions::community_id)))
+      .order_by(community::users_active_month.desc())
+      .select(community::id)
+      .limit(1)
+      .get_result::<CommunityId>(conn)
+      .await
+      .optional()?;
+
+    if let Some(largest_subscribed) = largest_subscribed {
+      // TODO Not sure this is correct
+      let posts = post::table
+        .filter(post::community_id.eq(largest_subscribed))
+        .limit(limit)
+        .select(Post::as_select())
+        .load::<Post>(conn)
+        .await?;
+
+      if (posts.len() as i64) < limit {
+        Ok(None)
+      } else {
+        let item = if page_back.unwrap_or_default() {
+          // for backward pagination, get first element instead
+          posts.into_iter().next()
+        } else {
+          posts.into_iter().next_back()
+        };
+        Ok(item)
+      }
+    } else {
+      // nothing subscribed to? no posts
+      Ok(None)
+    }
```

Comment:
I altered the prefetch to select from the tables directly, rather than have post_view list recursively call itself... but I'm not sure that these queries are correct.

cc @phiresky @dullbananas this should be looked at.